src/gt4py/storage/layout.py

```diff
@@ -0,0 +1,164 @@
+        res_layout = [0] * len(ranks)
+        for i, idx in enumerate(np.argsort(ranks)):
+            res_layout[idx] = i
+        return tuple(res_layout)
+
+    return layout_maker
+
+
+def layout_checker_factory(layout_maker):
+    def layout_checker(field: Union[np.ndarray, "cp.ndarray"], dimensions: Tuple[str, ...]) -> bool:
+        layout_map = layout_maker(dimensions)
+        return check_layout(layout_map, field.strides)
+
+    return layout_checker
+
+
+def _permute_layout_to_dimensions(
+    layout: Sequence[int], dimensions: Tuple[str, ...]
+) -> Tuple[int, ...]:
+    data_dims = [int(d) for d in dimensions if d.isdigit()]
+    canonical_dimensions = [d for d in "IJK" if d in dimensions] + [
+        str(d) for d in sorted(data_dims)
+    ]
+    res_layout = []
+    for d in dimensions:
+        res_layout.append(layout[canonical_dimensions.index(d)])
+    return tuple(res_layout)
+
+
+def make_gtcpu_kfirst_layout_map(dimensions: Tuple[str, ...]) -> Tuple[int, ...]:
+    layout = [i for i in range(len(dimensions))]
+    naxes = sum(dim in dimensions for dim in "IJK")
+    layout = [*layout[-naxes:], *layout[:-naxes]]
+    return _permute_layout_to_dimensions([lt for lt in layout if lt is not None], dimensions)
+
+
+def make_gtcpu_ifirst_layout_map(dimensions: Tuple[str, ...]) -> Tuple[int, ...]:
+    ctr = reversed(range(len(dimensions)))
+    layout = [next(ctr) for dim in "IJK" if dim in dimensions] + list(ctr)
+    if "K" in dimensions and "J" in dimensions:
+        if "I" in dimensions:
+            layout = [layout[0], layout[2], layout[1], *layout[3:]]
+        else:
+            layout = [layout[1], layout[0], *layout[2:]]
+    return _permute_layout_to_dimensions(layout, dimensions)
+
+
+def make_cuda_layout_map(dimensions: Tuple[str, ...]) -> Tuple[Optional[int], ...]:
+    layout = tuple(reversed(range(len(dimensions))))
+    return _permute_layout_to_dimensions(layout, dimensions)
+
+
+NaiveCPULayout: Final[LayoutInfo] = {
+    "alignment": 1,
+    "device": "cpu",
+    "layout_map": lambda axes: tuple(i for i in range(len(axes))),
+    "is_optimal_layout": lambda *_: True,
+}
+register("naive_cpu", NaiveCPULayout)
+
+CPUIFirstLayout: Final[LayoutInfo] = {
+    "alignment": 8,
+    "device": "cpu",
+    "layout_map": make_gtcpu_ifirst_layout_map,
+    "is_optimal_layout": layout_checker_factory(make_gtcpu_ifirst_layout_map),
+}
+register("cpu_ifirst", CPUIFirstLayout)
+
+
+CPUKFirstLayout: Final[LayoutInfo] = {
+    "alignment": 1,
+    "device": "cpu",
+    "layout_map": make_gtcpu_kfirst_layout_map,
+    "is_optimal_layout": layout_checker_factory(make_gtcpu_kfirst_layout_map),
+}
+register("cpu_kfirst", CPUKFirstLayout)
+
+
+CUDALayout: Final[LayoutInfo] = {
+    "alignment": 32,
+    "device": "gpu",
+    "layout_map": make_cuda_layout_map,
+    "is_optimal_layout": layout_checker_factory(make_cuda_layout_map),
+}
+register("cuda", CUDALayout)
+
+GPULayout: Final[LayoutInfo] = CUDALayout
+register("gpu", GPULayout)
```

Comment:
What's the idea/principle for only defining non-dace backends' layouts here while leaving the dace ones in-line withe backends?

Reply:
There are not super strong reasons, just that dace is an optional dependency and I find it slightly weird to define layouts for backends that might have not been installed.